python/ray/train/batch_predictor.py

```diff
@@ -170,7 +171,16 @@ def predict(
         ):
             predictor_kwargs["use_gpu"] = True
 
+        # For DLPredictor, we use numpy first for better dev API as well as
+        # improved performance of zero-copy data conversion
+        batch_format = "pandas"
         ctx = DatasetContext.get_current()
+        if issubclass(predictor_cls, DLPredictor):
+            # No casting or pandas conversion needed for numpy here, only
+            # at last output step
+            batch_format = "numpy"
+            ctx.enable_tensor_extension_casting = False
```

Comment:
Hmm is this needed/desired? I think that we'd still want to try to cast the tensor model outputs to our extension type, with a fallback?

If we are not going to do a transparent fallback, I don't think that we should mutate the context here (which would affect other future computations launched by this driver); we could instead just override `cast_tensor_columns` which is used on the model output below.

Reply:
I think this is less desirable for sure, more like last step i put up to ensure it works before going to sleep lol.

From my observation disabling all casting leads to worse perf in this numpy first path too (more disk spilling at least), so we definitely want to ensure it's enabled throughout our execution. I think we just need to ensure the fall back is there on the final pandas casting, ideally without mutating anything global. I can give a try to the `cast_tensor_columns` flag.

But my question is i think `cast_tensor_columns` also has longer duration of the entire execution rather than last step, maybe the best we can do is to just leave it as is, only by adding a try-catch at the end ?

Reply:
>I think we just need to ensure the fall back is there on the final pandas casting, ideally without mutating anything global. I can give a try to the cast_tensor_columns flag.

Agreed!

>But my question is i think cast_tensor_columns also has longer duration of the entire execution rather than last step, maybe the best we can do is to just leave it as is, only by adding a try-catch at the end ?

The local `cast_tensor_columns` variable (not the instance attribute) should only be applied to the model output cast below, AFAICT.

Reply:
We shouldn't disable tensor casting automatically. That must be done by the user only.

Reply:
Ah I see, we can leave the global and instance flag as is, only need to change `cast_tensor_columns` here for the numpy casting path

```
return convert_batch_type_to_pandas(
                    prediction_output, cast_tensor_columns
                )
```